src/kernels/raw/finder/localKernelFinder.node.ts

```diff
@@ -3,26 +3,182 @@
+    private cache: LocalKernelConnectionMetadata[] = [];
+    kind: string = 'local';
+
     constructor(
         @inject(LocalKnownPathKernelSpecFinder) private readonly nonPythonKernelFinder: LocalKnownPathKernelSpecFinder,
         @inject(LocalPythonAndRelatedNonPythonKernelSpecFinder)
-        private readonly pythonKernelFinder: LocalPythonAndRelatedNonPythonKernelSpecFinder
-    ) {}
+        private readonly pythonKernelFinder: LocalPythonAndRelatedNonPythonKernelSpecFinder,
+        @inject(IMemento) @named(GLOBAL_MEMENTO) private readonly globalState: Memento,
+        @inject(IFileSystem) private readonly fs: IFileSystem,
+        @inject(IApplicationEnvironment) private readonly env: IApplicationEnvironment,
+        @inject(IKernelFinder) kernelFinder: KernelFinder
+    ) {
+        kernelFinder.registerKernelFinder(this);
+    }
+
+    async activate(): Promise<void> {
+        noop();
+    }
+
+    async listContributedKernels(
+        resource: Resource,
+        cancelToken: CancellationToken | undefined,
+        useCache: 'ignoreCache' | 'useCache'
+    ): Promise<KernelConnectionMetadata[]> {
+        const kernels: KernelConnectionMetadata[] = await this.listKernelsImpl(resource, cancelToken, useCache).catch(
+            (ex) => {
+                // Sometimes we can get errors from the socket level or jupyter, with the message 'Canceled', lets ignore those
+                if (!isCancellationError(ex, true)) {
+                    traceError('Failed to get local kernels', ex);
+                }
+                return [];
+            }
+        );
+        traceVerbose(`KernelFinder discovered ${kernels.length} local`);
+
+        return kernels;
+    }
+
+    private async listKernelsImpl(
+        resource: Resource,
+        cancelToken: CancellationToken | undefined,
+        useCache: 'ignoreCache' | 'useCache'
+    ) {
+        const kernelsFromCachePromise =
+            useCache === 'ignoreCache' ? Promise.resolve([]) : this.getFromCache(cancelToken);
+        let updateCache = true;
+        const kernelsWithoutCachePromise: Promise<LocalKernelConnectionMetadata[]> = this.listKernels(
+            resource,
+            cancelToken
+        );
+        let kernels: LocalKernelConnectionMetadata[] = [];
+        if (useCache === 'ignoreCache') {
+            try {
+                kernels = await kernelsWithoutCachePromise;
+            } catch (ex) {
+                traceWarning(`Could not fetch kernels from the ${this.kind} server, falling back to cache: ${ex}`);
+                kernels = [];
+                updateCache = false;
+            }
+        } else {
+            let kernelsFromCache: LocalKernelConnectionMetadata[] | undefined;
+            kernelsFromCachePromise
+                .then((items) => {
+                    kernelsFromCache = items;
+                    updateCache = false;
+                })
+                .catch(noop);
```

Comment:
Looks like this would eat errors from getFromCache, cache would be less important, but I'd like to see something in the logging at least.

Reply:
This is originally from https://github.com/microsoft/vscode-jupyter/blob/main/src/kernels/kernelFinder.base.ts#L213